lib/src/conflicts.rs

```diff
@@ -276,37 +290,80 @@ fn write_conflict_marker(
     }
 }
 
-/// Parse a conflict marker from a line of a file. The conflict marker must have
-/// the correct length (CONFLICT_MARKER_LEN).
-fn parse_conflict_marker(line: &[u8]) -> Option<ConflictMarkerKind> {
+/// Parse a conflict marker from a line of a file. The conflict marker may have
+/// any length (even less than MIN_CONFLICT_MARKER_LEN).
+fn parse_conflict_marker_any_len(line: &[u8]) -> Option<ConflictMarker> {
     let first_byte = *line.first()?;
     let kind = ConflictMarkerKind::parse_byte(first_byte)?;
     let len = line.iter().take_while(|&&b| b == first_byte).count();
 
-    if len != CONFLICT_MARKER_LEN {
-        return None;
-    }
-
     if let Some(next_byte) = line.get(len) {
         // If there is a character after the marker, it must be ASCII whitespace
         if !next_byte.is_ascii_whitespace() {
             return None;
         }
     }
 
-    Some(kind)
+    Some(ConflictMarker { kind, len })
+}
+
+/// Parse a conflict marker, expecting it to have a certain length. Any conflict
+/// markers which have a different length are ignored.
+fn parse_conflict_marker(line: &[u8], expected_len: usize) -> Option<ConflictMarkerKind> {
+    debug_assert!(expected_len >= MIN_CONFLICT_MARKER_LEN);
+
+    parse_conflict_marker_any_len(line)
+        .filter(|marker| marker.len == expected_len)
```

Comment:
Maybe we need to accept `marker.len >= expected_len`? 

Suppose working-copy content was materialized with `N = 10`, and user removed a literal "<<<<<<<<<" (N = 9) without resolving any other conflicts, the new `expected_len` would drop. iirc, the working-copy content isn't updated to use the new `expected_len`.

Reply:
That's what I had originally, but I think it's too permissive actually. If we just had a minimum, and didn't enforce an exact number, then this could happen:

1. Conflict materialized with length 12 markers since file contains length 8 marker
2. User deletes length 8 marker
3. Conflict is parsed expecting length >=12 markers successfully
4. User adds length 8 marker back, since they made a mistake while editing
5. Conflict is parsed expecting length >=7 markers now, so both the length 8 and length 12 markers are parsed, even though the length 8 marker isn't a real marker

To solve this, I made it so that only the longest length markers in a file are parsed. So the example would look like this:

1. Conflict materialized with length 12 markers since file contains length 8 marker
2. User deletes length 8 marker
3. Conflict is parsed expecting length 12 markers successfully
4. User adds length 8 marker back, since they made a mistake while editing
5. Conflict is parsed expecting length 12 markers (since 12 is the longest marker present, and 12 >= 7), so the length 8 markers are correctly ignored

Reply:
Hmm, we can't know whether the re-added marker is a literal or conflict, right? My feeling is that it's safer to assume that a parsable marker is a conflict rather than invalidating any other conflicts by single addition of marker-looking line.

If we prefer stricter behavior, we'll probably need to store the materialized marker length to `TreeState` on checkout.

Reply:
Yeah, that's a good point. Sorry for the long response; I was just trying to think through all the reasonable options there might be and summarize them. I think these are the variables we can use for setting a bound:

* R = length of real conflict markers in original materialization (would need to be stored in `TreeState`)
* N = length of longest fake conflict markers in last snapshot's `Merge<FileId>`
* K = increment added to longest fake marker to get materialized marker length
* C = length of longest conflict marker (real or fake) in the raw file contents being snapshotted

I think it's important that we have a simple rule that the user can follow while editing a file to ensure that the real conflict markers will continue to be parsed successfully. These are the 4 options I can think of, along with the expressions that we could use to determine if a marker is parsable:

<details>
<summary>Possible options</summary>

**Option 1:** "Don't add any new fake markers which are longer than any fake markers in the last snapshotted version of the file"

Either `len >= max(7, N+K)` or `len >= max(7, N+1)` would work here.

**Option 2:** "Don't add any new fake markers which are longer than any fake markers in the original materialized file"

I think `len == max(7, N+K, C)` matches this rule, which is what I currently have implemented. The main difference with this and option 1 is that the user is allowed to add back fake conflict markers that they removed, and they still won't be parsed.

**Option 3:** "Don't add any new fake markers with length greater than or equal to the markers that `jj` added"

I think `len >= R` is the obvious implementation of this rule. It does require adding it to the `TreeState` though.

I think `len == max(7, N+1, C)` would also fall into this category, since the user could still add any fake markers that are shorter than the original materialized markers, since that would keep the original materialized markers as being the longest in the file, so it would continue parsing them successfully. This is different from `len == max(7, N+K, C)`, because that rule would cause snapshotting to fail the next time after adding any fake markers of length `N+1` through `N+K-1`, since that would result in the real conflict markers becoming considered shorter than allowed in the following snapshot.

**Option 4:** "Don't add any new fake markers with the exact same length as the markers that `jj` added"

This is probably the easiest rule to follow for the user, and it is what we currently have in `jj 0.23.0`, except that the materialized length is always 7. This could be achieved with `len == R`, which also requires adding it to the `TreeState`.

</details>

---

If we don't store the materialized marker length in the `TreeState`, then I believe adding a fake marker which is longer than the real materialized markers will **always** cause the real markers to not be parsed correctly. With some implementations it might fail to parse in the current snapshot (because it picked the wrong markers to parse), or with other implementations it might parse successfully in the current snapshot, but then fail in the next snapshot (because the inferred minimum conflict marker length increased as a result of there being new fake conflict markers added), or it could parse successfully but just have the wrong content (since some fake markers were treated as real ones and happened to form a valid conflict).

With that in mind, I feel like these are the 3 best options, each with different tradeoffs:

* `len >= max(7, N+1)`, since it is the simplest to implement, and it handles all of the common cases reasonably well.
* `len == max(7, N+1, C)`, since it allows adding any fake markers that are shorter than the real markers, which the user might expect to work.
* `len == R` since it relies on the fewest assumptions (but it requires storing marker length in the `TreeState`).

I think I'll look into adding it to the `TreeState` to see how difficult it would be. If it's a small change, then probably that would be the best option?

Reply:
> I think I'll look into adding it to the TreeState to see how difficult it would be. If it's a small change, then probably that would be the best option?

It wouldn't be difficult, but I'm not pretty sure if that's better because we're adding a new persistent state invisible to user. (BTW, the materialized conflict style could be stored in the TreeState if supported styles had ambiguous markers.)

@ilyagr, @martinvonz, any comments?

Reply:
> That's a good point about the invisible state. I think that's enough to convince me that one of the options is better.

Yeah, I agree. It would also make the behavior less predictable for merge tools, since the checkout state is even more hidden there (since the user might not even realize there's any state at all).

> Maybe another option is to use option 1 but additionally require that all markers have the same length, or otherwise interpret the file as resolved. I don't strongly which of these three options we go with.

I hadn't thought about that, that's a good idea. I think it feels unintuitive if we allow mixing different marker lengths within the same file, because it doesn't look like it should work visually. Based on that, I feel like my favorite options are currently:

* Parse any markers which are at least length `max(max_existing_marker_len + 1, MIN_CONFLICT_MARKER_LEN)`, but treat the file as resolved if there are multiple lengths of marker which could be parsed.
* Parse only the longest set of markers in the file which are at least length `max(max_existing_marker_len + 1, MIN_CONFLICT_MARKER_LEN)`, and ignore any other markers which are shorter.

Reply:
I prefer leaving conflict-looking content as conflict as possible. If the file were marked as resolved, user wouldn't notice that it had materialized conflict markers.

Reply:
True, that's also a good point. Does that mean you prefer Scott's option 1? I'm fine with that option.

Reply:
> I prefer leaving conflict-looking content as conflict as possible. If the file were marked as resolved, user wouldn't notice that it had materialized conflict markers.

If I'm understanding correctly, you're suggesting we allow mixing conflict markers sizes, and just parse any markers with length `>= expected_marker_len`, right?

My main issue with this right now is that I believe it can result in a situation where doing `touch <file>` will cause the file to change from conflicted to resolved on the next snapshot. Like if the current conflict markers are length 12, and somebody adds a `-------` conflict marker of length 15 outside of any existing conflict hunks, then it would be parsed correctly the first snapshot (since the `-------` line is outside of any hunk, it would be treated as plain text), but then after `touch <file>`, since the length 15 marker was added to the file in the working copy commit's tree, any conflict markers of length <= 15 will be ignored, so the file will be considered resolved.

Maybe this could be solved by marking the file as resolved immediately if any conflict marker lines are encountered outside of conflict hunks? Like if we see any other markers in the loop that searches for `<<<<<<<` and `>>>>>>>`, then we could mark the file as resolved. We would also probably have to mark JJ-style conflicts as resolved if they contain Git-style markers and vice versa.

Reply:
>  Like if the current conflict markers are length 12, and somebody adds a [...]

By the way, if you find time to add tests for cases like this one and the one you mentioned in https://github.com/martinvonz/jj/pull/4969#discussion_r1859717434, that would be very useful. Even if the behavior in those cases end up being something of a compromise, it's still useful to keep track of what the behavior is in a test case.

Reply:
> Does that mean you prefer Scott's option 1?

Yes. 1 or 3. I prefer 1 because it's also simpler to implement.

> My main issue with this right now is that I believe it can result in a situation where doing touch <file> will cause the file to change from conflicted to resolved on the next snapshot.

That's also a good point. I just assumed it would be less likely to insert longer conflict marker than editing conflict marker by mistake.

Reply:
> EDIT: btw, `touch <file>` should be safe because we check unmodified content explicitly (which I think can be replaced with content hash to be stored in file state.)

I tested it, and it's actually not safe with my current implementation, since the current check is based on the re-materialized conflict markers, and the re-materialized conflict markers would be longer than the actual conflict markers in the file, causing them not to match. I think using a content hash for the check would be a good idea to prevent this case at least.

---

Overall, I think these properties are all desirable, but we can only pick at most 2 of them (unless there's some other solution I'm missing):
1. There should be no hidden state (i.e. we shouldn't store marker length in `TreeState`)
2. Making a change that doesn't touch any conflict markers (e.g. `touch file` or `echo line >> file`) should never cause a file to become resolved if it is currently conflicted
3. Making a single conflict hunk invalid (e.g. wrong number of sides, invalid markers in hunk, etc.) shouldn't cause the file to become resolved if there are still other valid conflict hunks in the file

If we have 1 and 3, then we can't have 2, because when one hunk becomes resolved but still contains markers, it would cause the file to contain new markers, which would cause the existing markers to stop parsing on the next snapshot (e.g. after `touch file` or `echo line >> file`). If we have 1 and 2, then we can't have 3, since the way that we guarantee 2 would be by making parsing more strict, which sacrifices 3. We can get 2 and 3, but it would require sacrificing 1 by adding hidden state so that we know which conflict markers to parse always.

If this is the case, which two properties do we want to have? I'm personally leaning towards picking 1 and 2, since it seems like it gives the most predictable behavior to the user, but I can also see 2 and 3 being good because I'm not sure that hidden state would be an issue in practice. I personally would rather not sacrifice 2, since it seems like it would be surprising to the user that making a change unrelated to conflict markers could cause the file to become considered resolved.

Reply:
> * R = length of real conflict markers in original materialization (would need to be stored in `TreeState`)

Do we need to store it in the `TreeState`? Can we not get it from the previous content in the tree object? That might not help make the state visible to the user but it's nice if we avoid depending on `TreeState` for correctness.

Reply:
> > * R = length of real conflict markers in original materialization (would need to be stored in `TreeState`)
> 
> Do we need to store it in the `TreeState`? Can we not get it from the previous content in the tree object?

Do you mean getting it from the tree of the working-copy parent? It might work, but I'm not sure.

FWIW, I think it makes sense to trust `TreeState` here because the materialization parameter is kinda working-copy state, not a state of the committed tree.

Reply:
I updated the PR to first implement parsing if `>= materialized_marker_len`, and then I added `conflict_marker_len` to the file state in another commit to see how it could work. If we decide not to add it to the file state, I can always drop the last commit from the PR.

Reply:
> I was thinking that it's good if `TreeState` is just a cache that can we can discard without impact other than in the corner case where a file has been modified in a way that the cache prevents us from detecting (currently if the mtime and size is preserved but e.g. inode is changed).

Currently, I have it set to default to parsing length 7 conflict markers if there's no materialized conflict data in the `TreeState`, so this should still work in many cases since it will parse all conflict markers of length `>= 7` if the `TreeState` is reset, and the conflict markers should usually be length 7 anyways.

If we want to make it possible to reset the `TreeState` in more cases, we could have the conflict marker length default to `choose_materialized_conflict_marker_len(current_tree_merge)` instead maybe? There is a tradeoff though, because doing this could break already-materialized conflicts from older versions of `jj` if a user upgrades and the new materialized markers are longer than the actual markers in the file (which would be length 7 always in older versions).

Reply:
Sounds good to keep it simple for now.